streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -115,8 +160,168 @@ public void shutdown() throws IOException {
+        for (final String storeName : stateStores) {
+            for (final String key : keys) {
+                TestUtils.waitForCondition(new TestCondition() {
+                    @Override
+                    public boolean conditionMet() {
+                        final StreamsMetadata metadata = streams.metadataForKey(storeName, key, new StringSerializer());
+                        if (metadata == null) {
+                            return false;
+                        }
+                        final int index = metadata.hostInfo().port();
+                        final KafkaStreams streamsWithKey = streamRunnables[index].getStream();
+                        final ReadOnlyKeyValueStore<String, Long> store;
+                        try {
+                            store = streamsWithKey.store(storeName, QueryableStoreTypes.<String, Long>keyValueStore());
+                        } catch (IllegalStateException e) {
+                            // Kafka Streams instance may have closed but rebalance hasn't happened
+                            return false;
+                        }
+                        return store != null && store.get(key) != null;
+                    }
+                }, 30000, "waiting for metadata, store and value to be non null");
+            }
+        }
+    }
+
+    @Test
+    public void queryOnRebalance() throws Exception {
+        int numThreads = 2;
+        StreamRunnable[] streamRunnables = new StreamRunnable[numThreads];
+        Thread[] streamThreads = new Thread[numThreads];
+        final int numIterations = 500000;
+
+        // create concurrent producer
+        ProducerRunnable producerRunnable = new ProducerRunnable(STREAM_THREE, inputValues, numIterations);
+        Thread producerThread = new Thread(producerRunnable);
+
+        // create three stream threads
+        for (int i = 0; i < numThreads; i++) {
+            streamRunnables[i] = new StreamRunnable(STREAM_THREE, OUTPUT_TOPIC_THREE, i);
+            streamThreads[i] = new Thread(streamRunnables[i]);
+            streamThreads[i].start();
+        }
+        producerThread.start();
+
+        waitUntilAtLeastNumRecordProcessed(OUTPUT_TOPIC_THREE, 1);
+
+        for (int i = 0; i < numThreads; i++) {
+            verifyAllKVKeys(streamRunnables, streamRunnables[i].getStream(), inputValuesKeys,
+                new HashSet(Arrays.asList("word-count-store-" + STREAM_THREE)));
+        }
+
+        // kill N-1 threads
+        for (int i = 1; i < numThreads; i++) {
+            streamRunnables[i].close();
+            streamThreads[i].interrupt();
+            streamThreads[i].join();
+        }
+
+        // query from the remaining thread
+        verifyAllKVKeys(streamRunnables, streamRunnables[0].getStream(), inputValuesKeys,
+            new HashSet(Arrays.asList("word-count-store-" + STREAM_THREE)));
+        streamRunnables[0].close();
```

Comment:
Sorry @enothereska, you probably want to do the closing and joining of the threads either in the `shutdown` method or wrap the entire test in `try{...}finally{...}`. Otherwise the threads will still be hanging around when the test fails